### DISCUSSION TITLE
Batteries included

### DISCUSSION THREAD

**Discussion by JSG-0** [eProsima/Fast-DDS#4410] [Q&A]
Can there be made a single "batteries included" a.k.a. "kitchen sink" download that has all of the various (there are too many) repo packages  (foonathan whatever, dds gen, dds, cdr) in all the right places with **all** the required dependencies (gradle, java, cmake and their specific versions, etc). 

Please make the build simplier. IMO it takes an extreme and unecessary level of effort to make it happen to the point of avoiding updating anything. It is a real PITA because I also need to build for various architectures and build statically.

**Comment by JesusPoderoso** [comment]
Hi @JSG-0, thanks for using Fast DDS.
You are describing the [eProsima DDS Suite](https://github.com/eProsima/DDS-Suite). DDS Suite has all fixed versions per Fast DDS release (`fastdds-suite-v...`), and also LTS versions for a complete suite of eProsima products (`eprosima-dds-suite-v...`).
Hope it helps!

**Comment by JSG-0** [comment]
Thanks for the reply.
The documentation really is not *at all* clear to me:

  1.  That the “DDS Suite” is a batteries-included kitchen-sink all-inclusive. Four months ago I when I finally managed to build all the pieces it took 2 days, because I had to download, build and install very specific versions of the dependencies, all the fast dds repos, put them in various places (install specific versions of gradle and cmake, etc, etc, etc.).

So to be clear: are you saying that all this excess effort is no longer required to build everything? That would be great.





  1.  Is the docker image of the suite multi-platform? Does it support ubuntu on aarch64 *and* x86_64? Are the libraries static or only shared? If it is multiplatform and has static versions of the libs that would be perfect!

Thanks,
Jeffrey


From: Jesús Poderoso ***@***.***>
Sent: Tuesday, February 20, 2024 10:03 AM
To: eProsima/Fast-DDS ***@***.***>
Cc: GRAHAM, JEFFREY Collins ***@***.***>; Mention ***@***.***>
Subject: [External] Re: [eProsima/Fast-DDS] Batteries included (Discussion #4410)


Hi @JSG-0<https://urldefense.com/v3/__https:/github.com/JSG-0__;!!MvWE!DIHfO2Lk9jzmATn91cF1lFN21KHmhxjkOwduHGgPNUEmIoFey1rubvKmajMPacPsmArLt0tY4CkU03FLZg1mu3FHN-79uA$>, thanks for using Fast DDS.
You are describing the eProsima DDS Suite<https://urldefense.com/v3/__https:/github.com/eProsima/DDS-Suite__;!!MvWE!DIHfO2Lk9jzmATn91cF1lFN21KHmhxjkOwduHGgPNUEmIoFey1rubvKmajMPacPsmArLt0tY4CkU03FLZg1mu3FIFgdndg$>. DDS Suite has all fixed versions per Fast DDS release (fastdds-suite-v...), and also LTS versions for a complete suite of eProsima products (eprosima-dds-suite-v...).
Hope it helps!

**Comment by JSG-0** [comment]
How do I use the docker image to compile my own IDL and link my own static executable?
Can I use the docker image to cross-compile statically for aarch64?

From: Jesús Poderoso ***@***.***>
Sent: Tuesday, February 20, 2024 10:03 AM
To: eProsima/Fast-DDS ***@***.***>
Cc: GRAHAM, JEFFREY Collins ***@***.***>; Mention ***@***.***>
Subject: [External] Re: [eProsima/Fast-DDS] Batteries included (Discussion #4410)


Hi @JSG-0<https://urldefense.com/v3/__https:/github.com/JSG-0__;!!MvWE!DIHfO2Lk9jzmATn91cF1lFN21KHmhxjkOwduHGgPNUEmIoFey1rubvKmajMPacPsmArLt0tY4CkU03FLZg1mu3FHN-79uA$>, thanks for using Fast DDS.
You are describing the eProsima DDS Suite<https://urldefense.com/v3/__https:/github.com/eProsima/DDS-Suite__;!!MvWE!DIHfO2Lk9jzmATn91cF1lFN21KHmhxjkOwduHGgPNUEmIoFey1rubvKmajMPacPsmArLt0tY4CkU03FLZg1mu3FIFgdndg$>. DDS Suite has all fixed versions per Fast DDS release (fastdds-suite-v...), and also LTS versions for a complete suite of eProsima products (eprosima-dds-suite-v...).
Hope it helps!

**Comment by JesusPoderoso** [comment]
Hi Jeffrey,

> 1.  That the “DDS Suite” is a batteries-included kitchen-sink all-inclusive. Four months ago I when I finally managed to build all the pieces it took 2 days, because I had to download, build and install very specific versions of the dependencies, all the fast dds repos, put them in various places (install specific versions of gradle and cmake, etc, etc, etc.).
>
>        So to be clear: are you saying that all this excess effort is no longer required to build everything? That would be great.

DDS Suite keeps a complete track of Fast DDS versions and product-related versions, so it applies to run-time dependencies but not build-time dependencies. In such case, [Vulcanexus](https://vulcanexus.org/) may be the tool you are looking for.
There are available images on Dockerhub to download Vulcanexus over Ubuntu 22.04 for instance, with all the required dependencies already installed (cmake, java, etc).

> 2.  Is the docker image of the suite multi-platform? Does it support ubuntu on aarch64 *and* x86_64? Are the libraries static or only shared? If it is multiplatform and has static versions of the libs that would be perfect!

I don't think so, it only supports Ubuntu 22.04 amd64/x86_64, and all libraries are shared.
Thanks for your feedback, we will take it into account!

> How do I use the docker image to compile my own IDL and link my own static executable?

Using Fast DDS-Gen (`fastddsgen -help` command). It generates the type support per idl provided, based on the corresponding Fast DDS & Fast CDR versions. Your executable would need to import and use those generated types.

> Can I use the docker image to cross-compile statically for aarch64?

There should be toolchains for that task, but we don't provide them.
